### ISSUE TITLE
Support or warn for protocols with dunder methods and no trailing slash

### ISSUE THREAD

**Issue by philj56** [astral-sh/ty#2662]
### Summary

ty rejects matching protocols if a dunder method doesn't restrict itself to position-only arguments. This is fine, but the error message is rather hard to work out:

```python
from typing import Any, Protocol, runtime_checkable

@runtime_checkable
class WithSlash(Protocol):
    def __getitem__(self, key: Any, /) -> Any: ...

@runtime_checkable
class NoSlash(Protocol):
    def __getitem__(self, key: Any) -> Any: ...

a: WithSlash = {1: 2}  # No error
b: NoSlash = {1: 2}  # error[invalid-assignment]: Object of type `dict[Unknown | int, Unknown | int]` is not assignable to `NoSlash`
assert(isinstance(a, WithSlash))  # True
assert(isinstance(b, NoSlash))  # True
```
It'd be nice if ty (or perhaps ruff) checked for special dunder method declarations that do not match the necessary format, or had some other way of making this error clearer (or even accepting it as-is).

### Version

ty 0.0.14 (16597f5e2 2026-01-26)

**Comment by AlexWaygood** [comment]
> or had some other way of making this error clearer (or even accepting it as-is).

yes, we totally agree! This is pretty high-priority for us -- it's tracked in https://github.com/astral-sh/ty/issues/163. The only reason we haven't already done it is it will (unfortunately) probably be quite a large refactor

**Comment by carljm** [comment]
Will leave this open as a sub-issue since I think it's likely we may want to add an extra info sub-diagnostic for this specific case of callable assignability, beyond the general extra context that should be added in #163. Thanks for the report!

**Comment by carljm** [comment]
This issue is the generalization of #1644, which I mentioned in https://github.com/astral-sh/ty/issues/1644#issuecomment-3693137178

**Comment by sharkdp** [comment]
We now emit this error:

<img width="830" height="240" alt="Image" src="https://github.com/user-attachments/assets/3a4d27a4-9ae8-4d6e-a447-1cfabd2ecce2" />
